server/util.go

```diff
@@ -341,24 +343,30 @@ func (d *resultEncoder) columnTypeInfoCharsetID(info *ColumnInfo) uint16 {
 	return uint16(mysql.CharsetNameToID(d.chsName))
 }
 
+// encodeMeta encodes bytes for meta info like column names.
+// Note that the result should be consumed immediately.
 func (d *resultEncoder) encodeMeta(src []byte) []byte {
 	return d.encodeWith(src, d.encoding)
 }
 
+// encodeData encodes bytes for row data.
+// Note that the result should be consumed immediately.
 func (d *resultEncoder) encodeData(src []byte) []byte {
-	if d.isNull || d.isBinary {
+	if d.isNull || d.isBinary || d.dataIsBinary {
```

Comment:
This will cause some unexpect problems 😭, we should delete it. Like https://github.com/pingcap/tidb/issues/30690

Reply:
I find that the @@character_set_client is not considered when the column is binary(`blob`) type.

MySQL:
```sql
drop table if exists t;
create table t(a blob(100));
set character_set_connection = 'gbk';
set character_set_client = 'gbk';
set character_set_results = 'gbk';
insert into t values ("你好");
select hex(a) from t;
```
```
+--------------+
| hex(a)       |
+--------------+
| E4BDA0E5A5BD |
+--------------+
```

Reply:
Try this:
```
drop table if exists t;
create table t(a blob(100));
set character_set_client = 'gbk';
insert into t values ("你好");
select hex(a) from t;
```

MySQL will ignore the transform when `connection` and `client` are `gbk` which not same as TiDB does.

Reply:
Found you update your PR to solve blob problem, but it's not only for `blob`, try this:
```mysql
create table t (b blob, d json);
insert into t values ('你好', '{"测试": "你好"}');
select * from t;
```

So my question is why we need `dataIsBinary` here?

Reply:
Because the default value for `@@character_set_client` is `utf8mb4`. Note that the validation for utf8 encoding has been added, we need to identify which column is binary data and use `EncodingBinary` to skip this validation.

Reply:
This PR is too big, we can fix it in the next PR.
Please reopen the issue #30690 after this PR merged. @tangenta